node/core/pvf/src/execute/queue.rs

```diff
@@ -30,8 +30,21 @@ use futures::{
 	stream::{FuturesUnordered, StreamExt as _},
 	Future, FutureExt,
 };
+use polkadot_node_primitives::BACKING_EXECUTION_TIMEOUT;
+use polkadot_primitives::vstaging::{ExecutorParams, ExecutorParamsHash};
 use slotmap::HopSlotMap;
-use std::{collections::VecDeque, fmt, path::PathBuf, time::Duration};
+use std::{
+	collections::VecDeque,
+	fmt,
+	path::PathBuf,
+	time::{Duration, Instant},
+};
+
+/// The amount of time a job for which the queue does not have a compatible worker may wait in the
+/// queue. After that time passes, the queue will kill the first worker which becomes idle to
+/// re-spawn a new worker to execute the job immediately.
+const MAX_KEEP_WAITING: Duration =
+	Duration::from_millis(BACKING_EXECUTION_TIMEOUT.as_millis() as u64 * 2);
```

Comment:
I might be missing something but why did you choose this value? I would document the reasoning here. Also, isn't there a separate `APPROVAL_EXECUTION_TIMEOUT`? Why do we account only for the backing constant?

Reply:
I didn't get any constructive feedback about that value, so I had to decide for myself. `BACKING_EXECUTION_TIMEOUT` seems to be a fair base value as it is the minimal timeout for execution. Next, purely empirically, I feel it needs to be less than block time. And it still needs to be greater than a single execution timeout, otherwise, it loses its sense. So I've set it to `2 * BACKING_EXECUTION_TIMEOUT`. I would appreciate any better ideas.

Reply:
> And it still needs to be greater than a single execution timeout, otherwise, it loses its sense.

For backing execution I this is reasonable. But doesn't `APPROVAL_EXECUTION_TIMEOUT` lose its sense now, though? It is `12` seconds which is 3x the value being set here.

I don't have any better ideas, sorry. cc @eskimor @slumber

Reply:
It doesn't render `APPROVAL_EXECUTION_TIMEOUT` useless, as `MAX_KEEP_WAITING` is not a timeout by itself, it doesn't affect execution, it may only affect the order of execution. So if an approval candidate dangles in the queue for more than 4 secs because it has a different execution environment parameter set, after those 4 secs it still will be executed, it is just asked to wait for more tasks that can reuse existing workers to be executed before it.

Reply:
OK, in that case I think this value makes sense! Just remember to document the reasoning, please!

Just thinking here, do you reckon it would make sense to prioritize backing execution requests since the timeout is shorter?

Reply:
Ignore that last question for now, it can be a followup! Let's get this PR through! 😁

Reply:
Probably a really good and important idea! But requires some effort to put in (rn we don't know the type of job in the execution queue, is it backing or approval or something else). If @eskimor says it is a must-have, I'd instead set `MAX_KEEP_WAITING` to 0 right now and come out with a proper solution in a follow-up. The whole thing is already taking inexcusably much time :disappointed:

Reply:
backing actually should have the lowest priority. If there is a shortage on resources we have to make sure to be able to keep up with approvals and disputes. A slower backing phase is beneficial in that case.

For `MAX_KEEP_WAITING` ... value is in the right ball park. It is an edge case anyway, no need to worry about it too much.